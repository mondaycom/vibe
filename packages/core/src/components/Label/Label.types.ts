import { contentColors } from "../../utils/colors-vars-map";

export type LabelSizes = "small" | "medium";

export type LabelKind = "fill" | "line";

export type ContentColor = (typeof contentColors)[number];
export type LabelColor = "primary" | "dark" | "negative" | "positive" | ContentColor;
