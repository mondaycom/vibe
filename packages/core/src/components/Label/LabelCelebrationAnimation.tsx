import React, { cloneElement, forwardRef, useCallback, useRef, useState } from "react";
import cx from "classnames";
import useResizeObserver from "../../hooks/useResizeObserver";
import styles from "./LabelCelebrationAnimation.module.scss";

const DEFAULT_BORDER_RADIUS = 4;
const DEFAULT_STROKE_WIDTH = 1;

export interface LabelCelebrationAnimationProps {
  /**
   * The child component that will be wrapped by the animation.
   */
  children: React.ReactElement;
  /**
   * Callback fired when the celebration animation ends.
   */
  onAnimationEnd: () => void;
}

function LabelCelebrationAnimation({ children, onAnimationEnd }: LabelCelebrationAnimationProps) {
  const wrapperRef = useRef<HTMLDivElement>();
  const childRef = useRef<HTMLDivElement>();

  const [path, setPath] = useState<string>();

  const resizeObserverCallback = useCallback(
    ({ borderBoxSize }: { borderBoxSize: { blockSize: number; inlineSize: number } }) => {
      const { blockSize: height, inlineSize: width } = borderBoxSize || {};

      if (wrapperRef.current) {
        const d = getPath({ width, height });
        setPath(d);

        const perimeter = getPerimeter({ width, height });
        wrapperRef.current.style.setProperty("--container-perimeter", String(perimeter));
      }
    },
    []
  );

  useResizeObserver({
    ref: wrapperRef,
    callback: resizeObserverCallback,
    debounceTime: 0
  });

  const ChildComponentWithRef = forwardRef((_props, ref) =>
    cloneElement(children, {
      ref
    })
  );

  return (
    <div className={styles.celebration} ref={wrapperRef}>
      <svg className={styles.svg}>
        <path className={cx(styles.stroke, styles.base)} d={path} />
        <path className={cx(styles.stroke, styles.first)} d={path} />
        <path className={cx(styles.stroke, styles.second)} d={path} />
        <path className={cx(styles.stroke, styles.third)} d={path} onAnimationEnd={onAnimationEnd} />
      </svg>
      <ChildComponentWithRef ref={childRef} />
    </div>
  );
}

export default LabelCelebrationAnimation;

function getPath({
  width,
  height,
  borderRadius = DEFAULT_BORDER_RADIUS,
  strokeWidth = DEFAULT_STROKE_WIDTH
}: {
  width: number;
  height: number;
  borderRadius?: number;
  strokeWidth?: number;
}) {
  const offset = strokeWidth / 2;

  return `M ${width - strokeWidth / 2}, ${borderRadius} V ${
    height - borderRadius
  } A ${borderRadius} ${borderRadius} 0 0 1 ${width - borderRadius} ${height - strokeWidth / 2} H ${
    borderRadius + offset
  } A ${borderRadius} ${borderRadius} 0 0 1 ${strokeWidth / 2} ${height - borderRadius} V ${
    borderRadius + offset
  } A ${borderRadius} ${borderRadius} 0 0 1 ${borderRadius} ${strokeWidth / 2} L ${width - borderRadius}, ${
    strokeWidth / 2
  } A ${borderRadius} ${borderRadius} 0 0 1 ${width - strokeWidth / 2} ${borderRadius} Z`;
}

function getPerimeter({
  width,
  height,
  borderRadius = DEFAULT_BORDER_RADIUS
}: {
  width: number;
  height: number;
  borderRadius?: number;
}) {
  const straightWidth = width - 2 * borderRadius;
  const straightHeight = height - 2 * borderRadius;
  const cornerCircumference = 2 * Math.PI * borderRadius;
  return cornerCircumference + 2 * straightWidth + 2 * straightHeight;
}
