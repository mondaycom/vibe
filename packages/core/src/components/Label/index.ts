export { default as Label, type LabelProps } from "./Label";

export * from "./Label.types";
