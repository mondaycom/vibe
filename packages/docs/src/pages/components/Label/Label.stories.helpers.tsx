import React from "react";
import { StorybookLink, Tip } from "vibe-storybook-components";

export const TipCheckYourself = () => (
  <Tip title="Check yourself">
    Need to count or indicate numbers? Use the{" "}
    <StorybookLink page="Components/Counter" size={StorybookLink.sizes.SMALL}>
      Counter
    </StorybookLink>{" "}
    component instead.
  </Tip>
);
