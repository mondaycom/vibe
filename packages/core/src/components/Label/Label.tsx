import { camelCase } from "lodash-es";
import cx from "classnames";
import { ComponentDefaultTestId, getTestId } from "../../tests/test-ids-utils";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import React, { forwardRef, useCallback, useEffect, useMemo, useRef, useState } from "react";
import Text from "../Text/Text";
import Leg from "./Leg";
import { LabelAllowedColor as LabelColorEnum, LabelKind as LabelKindEnum, mapSizesToTextSize } from "./LabelConstants";
import { LabelColor, LabelKind, ContentColor } from "./Label.types";
import { contentColors } from "../../utils/colors-vars-map";
import { VibeComponentProps, withStaticProps } from "../../types";
import useClickableProps from "../../hooks/useClickableProps/useClickableProps";
import useMergeRef from "../../hooks/useMergeRef";
import styles from "./Label.module.scss";
import LabelCelebrationAnimation from "./LabelCelebrationAnimation";
import { LabelSizes } from "./Label.types";

export interface LabelProps extends VibeComponentProps {
  /**
   * Class name applied to the inner text wrapper.
   */
  labelClassName?: string;
  /**
   * The visual style of the label.
   */
  kind?: LabelKind;
  /**
   * The background color of the label.
   */
  color?: LabelColor;
  /**
   * The text content of the label.
   */
  text?: string;
  /**
   * If true, includes a leg (decorative extension).
   */
  isLegIncluded?: boolean;
  /**
   * Callback fired when the label is clicked.
   */
  onClick?: (event: React.MouseEvent<HTMLSpanElement, MouseEvent>) => void;
  /**
   * If true, triggers a celebration animation.
   */
  celebrationAnimation?: boolean;
  /**
   * The size of the label.
   */
  size?: LabelSizes;
}

const Label = forwardRef<HTMLElement, LabelProps>(
  (
    {
      className,
      labelClassName,
      kind = "fill",
      color = "primary",
      text = "",
      isLegIncluded = false,
      id,
      "data-testid": dataTestId,
      onClick,
      celebrationAnimation,
      size = "medium"
    }: LabelProps,
    ref
  ) => {
    const labelRef = useRef<HTMLSpanElement>(null);
    const mergedRef = useMergeRef(ref, labelRef);
    const [isCelebrationAnimation, setIsCelebrationAnimation] = useState(celebrationAnimation);

    const isClickable = Boolean(onClick);

    const classNames = useMemo(
      () =>
        cx(
          styles.label,
          getStyle(styles, camelCase("kind" + "-" + kind)),
          getStyle(styles, camelCase("color" + "-" + color)),
          {
            [styles.withLeg]: isLegIncluded,
            [styles.clickable]: isClickable,
            [styles.small]: size === "small"
          },
          labelClassName
        ),
      [kind, color, isLegIncluded, labelClassName, isClickable, size]
    );

    const backgroundColorStyle = useMemo(() => {
      if (contentColors.includes(color as ContentColor)) {
        return { backgroundColor: `var(--color-${color})` };
      }
    }, [color]);

    const onClickCallback = useCallback(
      (event: React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
        if (onClick) {
          event.preventDefault();
          onClick(event);
        }
      },
      [onClick]
    );

    const clickableProps = useClickableProps(
      {
        onClick: onClickCallback,
        id,
        ariaHidden: false,
        ariaHasPopup: false,
        ariaExpanded: false
      },
      labelRef
    );

    useEffect(() => {
      setIsCelebrationAnimation(celebrationAnimation);
    }, [celebrationAnimation]);

    const label = useMemo(() => {
      return (
        <span
          {...(isClickable && clickableProps)}
          className={cx({ [styles.clickableWrapper]: isClickable }, className)}
          data-testid={dataTestId || getTestId(ComponentDefaultTestId.LABEL, id)}
          ref={mergedRef}
        >
          <Text
            style={backgroundColorStyle}
            element="span"
            type={mapSizesToTextSize[size]}
            className={classNames}
            color="onInverted"
            data-celebration-text={isCelebrationAnimation}
          >
            <Text
              element="span"
              type={mapSizesToTextSize[size]}
              color="inherit"
              className={cx({ [styles.smallText]: size === "small" })}
            >
              {text}
            </Text>
            <span className={cx(styles.legWrapper)}>{isLegIncluded ? <Leg /> : null}</span>
          </Text>
        </span>
      );
    }, [
      isClickable,
      clickableProps,
      className,
      dataTestId,
      id,
      mergedRef,
      classNames,
      isCelebrationAnimation,
      text,
      isLegIncluded,
      size,
      backgroundColorStyle
    ]);

    // Celebration animation is applied only for line kind
    if (isCelebrationAnimation && kind === "line") {
      return (
        <LabelCelebrationAnimation onAnimationEnd={() => setIsCelebrationAnimation(false)}>
          {label}
        </LabelCelebrationAnimation>
      );
    }

    return label;
  }
);

interface LabelStaticProps {
  colors: typeof LabelColorEnum;
  kinds: typeof LabelKindEnum;
}

export default withStaticProps<LabelProps, LabelStaticProps>(Label, {
  colors: LabelColorEnum,
  kinds: LabelKindEnum
});
