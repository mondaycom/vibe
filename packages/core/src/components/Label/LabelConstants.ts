import { TextType } from "../Text/Text.types";
import Text from "../Text/Text";
import { LabelSizes } from "./Label.types";
import { ContentColorByName } from "../../utils/colors-vars-map";

/**
 * @deprecated
 */
export enum LabelKind {
  FILL = "fill",
  LINE = "line"
}

/**
 * @deprecated
 */
export enum LabelColor {
  PRIMARY = "primary",
  DARK = "dark",
  NEGATIVE = "negative",
  POSITIVE = "positive"
}

export const LabelAllowedColor = {
  ...LabelColor,
  ...ContentColorByName
};

export const mapSizesToTextSize: Record<LabelSizes, TextType> = {
  small: Text.types.TEXT3,
  medium: Text.types.TEXT2
};
