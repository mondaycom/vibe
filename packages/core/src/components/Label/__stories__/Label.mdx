import { Meta } from "@storybook/blocks";
import { Link } from "vibe-storybook-components";
import Label from "../Label";
import * as LabelStories from "./Label.stories";
import { TipCheckYourself } from "./Label.stories.helpers";

<Meta of={LabelStories} />

# Label

A label indicates the status of an item.

<Canvas of={LabelStories.Overview} />

### Import

```js
import { Label } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "Label will always appear in context next to the item it classifies.",
    "Displays a short message like new, beta, coming soon.,etc.",
    "Use only UI colors, not content colors (like status colors)."
  ]}
/>

<TipCheckYourself />

## Variants

### Kinds

<Canvas of={LabelStories.Kinds} />

### Size

Label can appear in 2 sizes: small and medium.

<Canvas of={LabelStories.Sizes} />

### Colors

<Canvas of={LabelStories.Colors} />

### Clickable

<Canvas of={LabelStories.Clickable} />

## Do’s and Don’ts

<ComponentRules
  rules={[
    {
      positive: {
        component: <Label text="New" />,
        description: "Use label to indicate the status of an item, for example: “New”."
      },
      negative: {
        component: <Label text="123" />,
        description: (
          <>
            Don’t use the label component in order to indicate numbers, instead use the{" "}
            <StorybookLink page="Components/Counter">Counter.</StorybookLink>
          </>
        )
      }
    },
    {
      positive: {
        component: (
          <div className="monday-storybook-label_inline-block">
            <h5 className="monday-storybook-label_title">Categories</h5>
            <Label text="new" />
          </div>
        ),
        description: "Use label only once per item."
      },
      negative: {
        component: [
          <div className="monday-storybook-label_block">
            <h5>Categories</h5>
            <div className="monday-storybook-label_inline-block">
              <Label text="Featured" />
              <Label text="CRM" />
              <Label text="Export" />
            </div>
          </div>
        ],
        description: (
          <>
            Don’t use multiple labels for one item. Instead, use{" "}
            <StorybookLink page="Components/Chips">Chips.</StorybookLink>
          </>
        )
      }
    },
    {
      positive: {
        component: [<Label text="New" kind="line" />, <Label text="Beta" />],
        description: "Use only the UI colors above."
      },
      negative: {
        component: [
          <Label className="monday-storybook-label_bad-lable monday-storybook-label_purple" kind="line" text="Beta" />,
          <Label className="monday-storybook-label_bad-lable monday-storybook-label_pink" kind="line" text="Beta" />
        ],
        description: "Don’t use any content colors for labels. If the page is full of CTAs, use the outline state."
      }
    }
  ]}
/>

## Use cases and examples

### Secondary label

In case of visual overload, use the secondary label in order to create hirarchy.

<Canvas of={LabelStories.SecondaryLabel} />

### Celebration

To celebrate new feature, outline label can be highlighted by adding celebrate animation.

<Canvas of={LabelStories.Celebration} />

## Related components

<RelatedComponents componentsNames={["Tooltip", "Counter", "Chips"]} />
