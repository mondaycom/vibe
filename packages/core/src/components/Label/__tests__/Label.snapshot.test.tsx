import { describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import Label from "../Label";

describe("Label renders correctly", () => {
  it("with empty props", () => {
    const tree = renderer.create(<Label />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("fill", () => {
    it("primary", () => {
      const tree = renderer.create(<Label text="Test" kind="fill" color="primary" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("dark", () => {
      const tree = renderer.create(<Label text="Test" kind="fill" color="dark" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("positive", () => {
      const tree = renderer.create(<Label text="Test" kind="fill" color="positive" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("negative", () => {
      const tree = renderer.create(<Label text="Test" kind="fill" color="negative" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("line", () => {
    it("primary", () => {
      const tree = renderer.create(<Label text="Test" kind="line" color="primary" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("dark", () => {
      const tree = renderer.create(<Label text="Test" kind="line" color="dark" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("positive", () => {
      const tree = renderer.create(<Label text="Test" kind="line" color="positive" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("negative", () => {
      const tree = renderer.create(<Label text="Test" kind="line" color="negative" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("without animation", () => {
      const tree = renderer.create(<Label text="Test" />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with leg", () => {
      const tree = renderer.create(<Label text="Test" isLegIncluded />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with className", () => {
      const tree = renderer.create(<Label className="test" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
