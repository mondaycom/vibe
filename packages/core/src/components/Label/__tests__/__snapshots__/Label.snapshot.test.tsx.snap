// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label renders correctly fill dark 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindFill colorDark"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      Test
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;

exports[`Label renders correctly fill negative 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindFill colorNegative"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      Test
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;

exports[`Label renders correctly fill positive 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindFill colorPositive"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      Test
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;

exports[`Label renders correctly fill primary 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindFill colorPrimary"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      Test
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;

exports[`Label renders correctly line dark 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindLine colorDark"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      Test
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;

exports[`Label renders correctly line negative 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindLine colorNegative"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      Test
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;

exports[`Label renders correctly line positive 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindLine colorPositive"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      Test
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;

exports[`Label renders correctly line primary 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindLine colorPrimary"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      Test
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;

exports[`Label renders correctly line with className 1`] = `
<span
  className="test"
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindFill colorPrimary"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;

exports[`Label renders correctly line with leg 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindFill colorPrimary withLeg"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      Test
    </span>
    <span
      className="legWrapper"
    >
      <svg
        fill="none"
        height="5"
        viewBox="0 0 8 5"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0V2.76772C0 4.21783 1.77351 4.92116 2.76736 3.86519L5.45681 1.00765C6.06097 0.365738 6.90294 0.00127805 7.7843 3.35493e-06C5.30942 2.53165e-06 2.69058 1.4328e-06 0 0Z"
          fill="#E44258"
        />
      </svg>
    </span>
  </span>
</span>
`;

exports[`Label renders correctly line without animation 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindFill colorPrimary"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      Test
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;

exports[`Label renders correctly with empty props 1`] = `
<span
  className=""
  data-testid="label"
>
  <span
    className="typography onInverted start singleLineEllipsis text text2Normal label kindFill colorPrimary"
    data-testid="text"
  >
    <span
      className="typography inherit start singleLineEllipsis text text2Normal"
      data-testid="text"
    >
      
    </span>
    <span
      className="legWrapper"
    />
  </span>
</span>
`;
